File: src/main/java/imgui/dsl_.java
package imgui;

import glm_.vec2.Vec2;
import glm_.vec4.Vec4;

/**
 * twin brother of dsl, manual overloads
 */
public class dsl_ {

    // Windows

    public static void window(String name, Runnable block) {
        window(name, null, 0, block);
    }

    public static void window(String name, MutableProperty0<Boolean> open, Runnable block) {
        window(name, open, 0, block);
    }

    public static void window(String name, MutableProperty0<Boolean> open, int windowFlags, Runnable block) {
        if (ImGui.INSTANCE.begin(name, open, windowFlags)) // ~open
            block.run();
        ImGui.INSTANCE.end();
    }

    // Child Windows

    public static void child(String strId, Runnable block) {
        child(strId, new Vec2(), false, 0, block);
    }

    public static void child(String strId, Vec2 size, Runnable block) {
        child(strId, size, false, 0, block);
    }

    public static void child(String strId, Vec2 size, boolean border, Runnable block) {
        child(strId, size, border, 0, block);
    }

    public static void child(String strId, Vec2 size, boolean border, int windowFlags, Runnable block) {
        if (ImGui.INSTANCE.beginChild(strId, size, border, windowFlags)) // ~open
            block.run();
        ImGui.INSTANCE.endChild();
    }

    // Parameters stacks (shared)

    public static void withFont(Runnable block) {
        withFont(ImGui.INSTANCE.getDefaultFont(), block);
    }

    public static void withFont(Font font, Runnable block) {
        ImGui.INSTANCE.pushFont(font);
        block.run();
        ImGui.INSTANCE.popFont();
    }

    private static void _push(Col idx, Object col) {
        if (col instanceof Integer)
            ImGui.INSTANCE.pushStyleColor(idx, (int) col);
        else
            ImGui.INSTANCE.pushStyleColor(idx, (Vec4) col);
    }

    public static void withStyleColor(Col idx, Object col, Runnable block) {
        _push(idx, col);
        block.run();
        ImGui.INSTANCE.popStyleColor(1);
    }

    public static void withStyleColor(Col idx0, Object col0,
                                      Col idx1, Object col1, Runnable block) {
        _push(idx0, col0);
        _push(idx1, col1);
        block.run();
        ImGui.INSTANCE.popStyleColor(2);
    }

    public static void withStyleColor(Col idx0, Object col0,
                                      Col idx1, Object col1,
                                      Col idx2, Object col2, Runnable block) {
        _push(idx0, col0);
        _push(idx1, col1);
        _push(idx2, col2);
        block.run();
        ImGui.INSTANCE.popStyleColor(3);
    }

    public static void withStyleColor(Col idx0, Object col0,
                                      Col idx1, Object col1,
                                      Col idx2, Object col2,
                                      Col idx3, Object col3, Runnable block) {
        _push(idx0, col0);
        _push(idx1, col1);
        _push(idx2, col2);
        _push(idx3, col3);
        block.run();
        ImGui.INSTANCE.popStyleColor(4);
    }

    public static void withStyleColor(Col idx0, Object col0,
                                      Col idx1, Object col1,
                                      Col idx2, Object col2,
                                      Col idx3, Object col3,
                                      Col idx4, Object col4, Runnable block) {
        _push(idx0, col0);
        _push(idx1, col1);
        _push(idx2, col2);
        _push(idx3, col3);
        _push(idx4, col4);
        block.run();
        ImGui.INSTANCE.popStyleColor(5);
    }

    public static void withStyleVar(StyleVar idx, Object value, Runnable block) {
        ImGui.INSTANCE.pushStyleVar(idx, value);
        block.run();
        ImGui.INSTANCE.popStyleVar(1);
    }

    // Parameters stacks (current window)

    public static void withItemWidth(int itemWidth, Runnable block) {
        withItemWidth((float) itemWidth, block);
    }

    public static void withItemWidth(float itemWidth, Runnable block) {
        ImGui.INSTANCE.pushItemWidth(itemWidth);
        block.run();
        ImGui.INSTANCE.popItemWidth();
    }

    public static void withTextWrapPos(Runnable block) {
        withTextWrapPos(0f, block);
    }

    public static void withTextWrapPos(float wrapPosX, Runnable block) {
        ImGui.INSTANCE.pushTextWrapPos(wrapPosX);
        block.run();
        ImGui.INSTANCE.popTextWrapPos();
    }

    public static void withAllowKeyboardFocus(boolean allowKeyboardFocus, Runnable block) {
        ImGui.INSTANCE.pushAllowKeyboardFocus(allowKeyboardFocus);
        block.run();
        ImGui.INSTANCE.popAllowKeyboardFocus();
    }

    public static void withButtonRepeat(boolean repeat, Runnable block) {
        ImGui.INSTANCE.pushButtonRepeat(repeat);
        block.run();
        ImGui.INSTANCE.popButtonRepeat();
    }


    // Cursor / Layout

    public static void indent(Runnable block) {
        indent(0f, block);
    }

    public static void indent(float indentW, Runnable block) {
        ImGui.INSTANCE.indent(indentW);
        block.run();
        ImGui.INSTANCE.unindent(indentW);
    }

    public static void group(Runnable block) {
        ImGui.INSTANCE.beginGroup();
        block.run();
        ImGui.INSTANCE.endGroup();
    }


    // ID stack/scopes

    public static void withId(int id, Runnable block) {
        ImGui.INSTANCE.pushId(id);
        block.run();
        ImGui.INSTANCE.popId();
    }

    public static void withId(String id, Runnable block) {
        ImGui.INSTANCE.pushId(id);
        block.run();
        ImGui.INSTANCE.popId();
    }


    // Widgets: Main

    public static void button(String label, Runnable block) {
        button(label, new Vec2(), block);
    }

    public static void button(String label, Vec2 sizeArg, Runnable block) {
        if (ImGui.INSTANCE.button(label, sizeArg))
            block.run();
    }

    public static void smallButton(String label, Runnable block) {
        if (ImGui.INSTANCE.smallButton(label))
            block.run();
    }

    public static void invisibleButton(String strId, Vec2 sizeArg, Runnable block) {
        if (ImGui.INSTANCE.invisibleButton(strId, sizeArg))
            block.run();
    }

    public static void arrowButton(String id, Dir dir, Runnable block) {
        if (ImGui.INSTANCE.arrowButton(id, dir))
            block.run();
    }

    public static void imageButton(int userTextureId, Vec2 size, Vec4 bgCol, Runnable block) {
        imageButton(userTextureId, size, new Vec2(), new Vec2(), -1, bgCol, new Vec4(1), block);
    }

    public static void imageButton(int userTextureId, Vec2 size, Vec2 uv0, Vec2 uv1, Vec4 bgCol, Runnable block) {
        imageButton(userTextureId, size, uv0, uv1, -1, bgCol, new Vec4(1), block);
    }

    public static void imageButton(int userTextureId, Vec2 size, Vec2 uv0, Vec2 uv1, int framePadding, Vec4 bgCol, Runnable block) {
        imageButton(userTextureId, size, uv0, uv1, framePadding, bgCol, new Vec4(1), block);
    }

    public static void imageButton(int userTextureId, Vec2 size, Vec2 uv0, Vec2 uv1, int framePadding, Vec4 bgCol, Vec4 tintCol, Runnable block) {
        if (ImGui.INSTANCE.imageButton(userTextureId, size, uv0, uv1, framePadding, bgCol, tintCol))
            block.run();
    }

    public static void checkbox(String label, MutableProperty0<Boolean> vPtr, Runnable block) {
        if (ImGui.INSTANCE.checkbox(label, vPtr))
            block.run();
    }

    public static void checkboxFlags(String label, MutableProperty0<Integer> vPtr, int flagsValue, Runnable block) {
        if (ImGui.INSTANCE.checkboxFlags(label, vPtr, flagsValue))
            block.run();
    }

    public static void radioButton(String label, boolean active, Runnable block) {
        if (ImGui.INSTANCE.radioButton(label, active))
            block.run();
    }

    public static void radioButton(String label, MutableProperty0<Integer> v, int vButton, Runnable block) {
        if (ImGui.INSTANCE.radioButton(label, v, vButton))
            block.run();
    }


    // Widgets: Combo Box


    public static void useCombo(String label, String previewValue, Runnable block) {
        useCombo(label, previewValue, 0, block);
    }

    public static void useCombo(String label, String previewValue, int comboFlags, Runnable block) {
        if (ImGui.INSTANCE.beginCombo(label, previewValue, comboFlags))
            block.run();
        ImGui.INSTANCE.endCombo();
    }

    public static void combo(String label, MutableProperty0<Integer> currentItem, String itemsSeparatedByZeros, Runnable block) {
        combo(label, currentItem, itemsSeparatedByZeros, -1, block);
    }

    public static void combo(String label, MutableProperty0<Integer> currentItem, String itemsSeparatedByZeros, int heightInItems, Runnable block) {
        if (ImGui.INSTANCE.combo(label, currentItem, itemsSeparatedByZeros, heightInItems))
            block.run();
    }


    // Widgets: Trees

    public static void treeNode(String label, Runnable block) {
        if (ImGui.INSTANCE.treeNode(label)) {
            block.run();
            ImGui.INSTANCE.treePop();
        }
    }

    public static void treeNode(String strId, String fmt, Runnable block) {
        if (ImGui.INSTANCE.treeNode(strId, fmt)) {
            block.run();
            ImGui.INSTANCE.treePop();
        }
    }

    public static void treeNode(Object ptrId, String fmt, Runnable block) {
        if (ImGui.INSTANCE.treeNode(ptrId, fmt)) {
            block.run();
            ImGui.INSTANCE.treePop();
        }
    }

    public static void treePushed(Object ptrId, Runnable block) {
        ImGui.INSTANCE.treePush(ptrId);
        ImGui.INSTANCE.treePop();
    }

    public static void collapsingHeader(String label, Runnable block) {
        collapsingHeader(label, 0, block);
    }

    public static void collapsingHeader(String label, int treeNodeFlags, Runnable block) {
        if (ImGui.INSTANCE.collapsingHeader(label, treeNodeFlags))
            block.run();
    }

    public static void collapsingHeader(String label, MutableProperty0<Boolean> open, Runnable block) {
        collapsingHeader(label, open, 0, block);
    }

    public static void collapsingHeader(String label, MutableProperty0<Boolean> open, int treeNodeFlags, Runnable block) {
        if (ImGui.INSTANCE.collapsingHeader(label, open, treeNodeFlags))
            block.run();
    }


    // Widgets: Selectables

    public static void selectable(String label, Runnable block) {
        selectable(label, false, 0, new Vec2(), block);
    }

    public static void selectable(String label, boolean selected, Runnable block) {
        selectable(label, selected, 0, new Vec2(), block);
    }

    public static void selectable(String label, boolean selected, int flags, Runnable block) {
        selectable(label, selected, flags, new Vec2(), block);
    }

    public static void selectable(String label, boolean selected, int flags, Vec2 sizeArg, Runnable block) {
        if (ImGui.INSTANCE.selectable(label, selected, flags, sizeArg))
            block.run();
    }


    // Widgets: Menus

    public static void mainMenuBar(Runnable block) {
        if (ImGui.INSTANCE.beginMainMenuBar()) {
            block.run();
            ImGui.INSTANCE.endMainMenuBar();
        }
    }

    public static void menuBar(Runnable block) {
        if (ImGui.INSTANCE.beginMenuBar()) {
            block.run();
            ImGui.INSTANCE.endMenuBar();
        }
    }

    public static void menu(String label, Runnable block) {
        menu(label, true, block);
    }

    public static void menu(String label, boolean enabled, Runnable block) {
        if (ImGui.INSTANCE.beginMenu(label, enabled)) {
            block.run();
            ImGui.INSTANCE.endMenu();
        }
    }

    public static void menuItem(String label, Runnable block) {
        menuItem(label, "", false, true, block);
    }

    public static void menuItem(String label, String shortcut, Runnable block) {
        menuItem(label, shortcut, false, true, block);
    }

    public static void menuItem(String label, String shortcut, boolean selected, Runnable block) {
        menuItem(label, shortcut, selected, true, block);
    }

    public static void menuItem(String label, String shortcut, boolean selected, boolean enabled, Runnable block) {
        if (ImGui.INSTANCE.menuItem(label, shortcut, selected, enabled))
            block.run();
    }


    // Tooltips

    public static void tooltip(Runnable block) {
        ImGui.INSTANCE.beginTooltip();
        block.run();
        ImGui.INSTANCE.endTooltip();
    }


    // Popups, Modals

    public static void popup(String strId, Runnable block) {
        popup(strId, 0, block);
    }

    public static void popup(String strId, int windowFlags, Runnable block) {
        if (ImGui.INSTANCE.beginPopup(strId, windowFlags)) {
            block.run();
            ImGui.INSTANCE.endPopup();
        }
    }

    public static void popupContextItem(String strId, Runnable block) {
        popupContextItem(strId, 1, block);
    }

    public static void popupContextItem(String strId, int mouseButton, Runnable block) {
        if (ImGui.INSTANCE.beginPopupContextItem(strId, mouseButton)) {
            block.run();
            ImGui.INSTANCE.endPopup();
        }
    }

    public static void popupContextWindow(String strId, Runnable block) {
        popupContextWindow(strId, 1, true, block);
    }

    public static void popupContextWindow(String strId, int mouseButton, Runnable block) {
        popupContextWindow(strId, mouseButton, true, block);
    }

    public static void popupContextWindow(String strId, int mouseButton, boolean alsoOverItems, Runnable block) {
        if (ImGui.INSTANCE.beginPopupContextWindow(strId, mouseButton, alsoOverItems)) {
            block.run();
            ImGui.INSTANCE.endPopup();
        }
    }

    public static void popupContextVoid(String strId, Runnable block) {
        popupContextVoid(strId, 1, block);
    }

    public static void popupContextVoid(String strId, int mouseButton, Runnable block) {
        if (ImGui.INSTANCE.beginPopupContextVoid(strId, mouseButton)) {
            block.run();
            ImGui.INSTANCE.endPopup();
        }
    }

    public static void popupModal(String name, Runnable block) {
        popupModal(name, null, 0, block);
    }

    public static void popupModal(String name, MutableProperty0<Boolean> pOpen, Runnable block) {
        popupModal(name, pOpen, 0, block);
    }

    public static void popupModal(String name, MutableProperty0<Boolean> pOpen, int windowFlags, Runnable block) {
        if (ImGui.INSTANCE.beginPopupModal(name, pOpen, windowFlags)) {
            block.run();
            ImGui.INSTANCE.endPopup();
        }
    }


    // Tab Bars, Tabs

    public static void tabBar(String strId, Runnable block) {
        tabBar(strId, 0, block);
    }

    public static void tabBar(String strId, int tabBarFlags, Runnable block) {
        if (ImGui.INSTANCE.beginTabBar(strId, tabBarFlags))
            block.run();
        ImGui.INSTANCE.endTabBar();
    }

    public static void tabItem(String label, Runnable block) {
        tabItem(label, null, 0, block);
    }

    public static void tabItem(String label, MutableProperty0<Boolean> pOpen, Runnable block) {
        tabItem(label, pOpen, 0, block);
    }

    public static void tabItem(String label, MutableProperty0<Boolean> pOpen, int tabItemFlags, Runnable block) {
        if (ImGui.INSTANCE.beginTabItem(label, pOpen, tabItemFlags))
            block.run();
        ImGui.INSTANCE.endTabItem();
    }


    // Drag and Drop

    public static void dragDropSource(Runnable block) {
        dragDropSource(0, block);
    }

    public static void dragDropSource(int dragDropFlags, Runnable block) {
        if (ImGui.INSTANCE.beginDragDropSource(dragDropFlags)) {
            block.run();
            ImGui.INSTANCE.endDragDropSource();
        }
    }

    public static void dragDropTarget(Runnable block) {
        if (ImGui.INSTANCE.beginDragDropTarget()) {
            block.run();
            ImGui.INSTANCE.endDragDropTarget();
        }
    }


    // Clipping

    public static void withClipRect(Vec2 clipRectMin, Vec2 clipRectMax, boolean intersectWithCurrentClipRect, Runnable block) {
        ImGui.INSTANCE.pushClipRect(clipRectMin, clipRectMax, intersectWithCurrentClipRect);
        block.run();
        ImGui.INSTANCE.popClipRect();
    }


    // Miscellaneous Utilities

    public static void childFrame(int id, Vec2 size, Runnable block) {
        childFrame(id, size, 0, block);
    }

    public static void childFrame(int id, Vec2 size, int windowFlags, Runnable block) {
        ImGui.INSTANCE.beginChildFrame(id, size, windowFlags);
        block.run();
        ImGui.INSTANCE.endChildFrame();
    }
}

File: src/main/kotlin/imgui/dsl.kt
package imgui

import glm_.f
import glm_.vec2.Vec2
import glm_.vec4.Vec4
import imgui.ImGui.begin
import imgui.ImGui.beginChild
import imgui.ImGui.beginGroup
import imgui.ImGui.beginMainMenuBar
import imgui.ImGui.beginMenu
import imgui.ImGui.beginMenuBar
import imgui.ImGui.beginPopup
import imgui.ImGui.beginPopupContextItem
import imgui.ImGui.beginPopupContextWindow
import imgui.ImGui.beginPopupModal
import imgui.ImGui.beginTooltip
import imgui.ImGui.collapsingHeader
import imgui.ImGui.combo
import imgui.ImGui.defaultFont
import imgui.ImGui.end
import imgui.ImGui.endChild
import imgui.ImGui.endGroup
import imgui.ImGui.endMainMenuBar
import imgui.ImGui.endMenu
import imgui.ImGui.endMenuBar
import imgui.ImGui.endPopup
import imgui.ImGui.endTooltip
import imgui.ImGui.indent
import imgui.ImGui.menuItem
import imgui.ImGui.popAllowKeyboardFocus
import imgui.ImGui.popButtonRepeat
import imgui.ImGui.popFont
import imgui.ImGui.popId
import imgui.ImGui.popItemWidth
import imgui.ImGui.popStyleColor
import imgui.ImGui.popStyleVar
import imgui.ImGui.popTextWrapPos
import imgui.ImGui.pushAllowKeyboardFocus
import imgui.ImGui.pushButtonRepeat
import imgui.ImGui.pushFont
import imgui.ImGui.pushId
import imgui.ImGui.pushItemWidth
import imgui.ImGui.pushStyleColor
import imgui.ImGui.pushStyleVar
import imgui.ImGui.pushTextWrapPos
import imgui.ImGui.selectable
import imgui.ImGui.treeNode
import imgui.ImGui.treePop
import imgui.ImGui.unindent
import kotlin.reflect.KMutableProperty0

object dsl {

    // Windows

    inline fun window(name: String, open: KMutableProperty0<Boolean>? = null, flags: WindowFlags = 0, block: () -> Unit) {
        if (begin(name, open, flags)) // ~open
            block()
        end()
    }

    // Child Windows

    inline fun child(strId: String, size: Vec2 = Vec2(), border: Boolean = false, extraFlags: WindowFlags = 0, block: () -> Unit) {
        if (beginChild(strId, size, border, extraFlags)) // ~open
            block()
        endChild()
    }

    // Parameters stacks (shared)

    inline fun withFont(font: Font = defaultFont, block: () -> Unit) {
        pushFont(font)
        block()
        popFont()
    }

    fun _push(idx: Col, col: Any) {
        if (col is Int)
            pushStyleColor(idx, col)
        else
            pushStyleColor(idx, col as Vec4)
    }

    inline fun withStyleColor(idx: Col, col: Any, block: () -> Unit) {
        _push(idx, col)
        block()
        popStyleColor()
    }

    inline fun withStyleColor(idx0: Col, col0: Any,
                              idx1: Col, col1: Any, block: () -> Unit) {
        _push(idx0, col0)
        _push(idx1, col1)
        block()
        popStyleColor(2)
    }

    inline fun withStyleColor(idx0: Col, col0: Any,
                              idx1: Col, col1: Any,
                              idx2: Col, col2: Any, block: () -> Unit) {
        _push(idx0, col0)
        _push(idx1, col1)
        _push(idx2, col2)
        block()
        popStyleColor(3)
    }

    inline fun withStyleColor(idx0: Col, col0: Any,
                              idx1: Col, col1: Any,
                              idx2: Col, col2: Any,
                              idx3: Col, col3: Any,
                              block: () -> Unit) {
        _push(idx0, col0)
        _push(idx1, col1)
        _push(idx2, col2)
        _push(idx3, col3)
        block()
        popStyleColor(4)
    }

    inline fun withStyleColor(idx0: Col, col0: Any,
                              idx1: Col, col1: Any,
                              idx2: Col, col2: Any,
                              idx3: Col, col3: Any,
                              idx4: Col, col4: Any, block: () -> Unit) {
        _push(idx0, col0)
        _push(idx1, col1)
        _push(idx2, col2)
        _push(idx3, col3)
        _push(idx4, col4)
        block()
        popStyleColor(5)
    }

    inline fun withStyleVar(idx: StyleVar, value: Any, block: () -> Unit) {
        pushStyleVar(idx, value)
        block()
        popStyleVar()
    }

    // Parameters stacks (current window)

    inline fun withItemWidth(itemWidth: Int, block: () -> Unit) = withItemWidth(itemWidth.f, block)
    inline fun withItemWidth(itemWidth: Float, block: () -> Unit) {
        pushItemWidth(itemWidth)
        block()
        popItemWidth()
    }

    inline fun withTextWrapPos(wrapPosX: Float = 0f, block: () -> Unit) {
        pushTextWrapPos(wrapPosX)
        block()
        popTextWrapPos()
    }

    inline fun withPushAllowKeyboardFocus(allowKeyboardFocus: Boolean, block: () -> Unit) {
        pushAllowKeyboardFocus(allowKeyboardFocus)
        block()
        popAllowKeyboardFocus()
    }

    inline fun withPushButtonRepeat(repeat: Boolean, block: () -> Unit) {
        pushButtonRepeat(repeat)
        block()
        popButtonRepeat()
    }


    // Cursor / Layout

    inline fun withIndent(indentW: Float = 0f, block: () -> Unit) {
        indent(indentW)
        block()
        unindent(indentW)
    }

    inline fun withGroup(block: () -> Unit) {
        beginGroup()
        block()
        endGroup()
    }


    // ID stack/scopes

    inline fun withId(id: Int, block: () -> Unit) {
        pushId(id)
        block()
        popId()
    }

    inline fun withId(id: String, block: () -> Unit) {
        pushId(id)
        block()
        popId()
    }


    // Widgets: Main

    inline fun button(label: String, sizeArg: Vec2 = Vec2(), block: () -> Unit) {
        if (ImGui.button(label, sizeArg))
            block()
    }

    inline fun smallButton(label: String, block: () -> Unit) {
        if (ImGui.smallButton(label))
            block()
    }

    inline fun invisibleButton(strId: String, sizeArg: Vec2, block: () -> Unit) {
        if (ImGui.invisibleButton(strId, sizeArg))
            block()
    }

    inline fun arrowButton(id: String, dir: Dir, block: () -> Unit) {
        if (ImGui.arrowButton(id, dir))
            block()
    }

    inline fun imageButton(userTextureId: TextureID, size: Vec2, uv0: Vec2 = Vec2(), uv1: Vec2 = Vec2(),
                           framePadding: Int = -1, bgCol: Vec4 = Vec4(), tintCol: Vec4 = Vec4(1), block: () -> Unit) {
        if (ImGui.imageButton(userTextureId, size, uv0, uv1, framePadding, bgCol, tintCol))
            block()
    }

    inline fun checkbox(label: String, vPtr: KMutableProperty0<Boolean>, block: () -> Unit) {
        if (ImGui.checkbox(label, vPtr))
            block()
    }

    inline fun checkboxFlags(label: String, vPtr: KMutableProperty0<Int>, flagsValue: Int, block: () -> Unit) {
        if (ImGui.checkboxFlags(label, vPtr, flagsValue))
            block()
    }

    inline fun radioButton(label: String, active: Boolean, block: () -> Unit) {
        if (ImGui.radioButton(label, active))
            block()
    }

    inline fun radioButton(label: String, v: KMutableProperty0<Int>, vButton: Int, block: () -> Unit) {
        if (ImGui.radioButton(label, v, vButton))
            block()
    }


    // Widgets: Combo Box


    inline fun useCombo(label: String, previewValue: String?, flags: ComboFlags = 0, block: () -> Unit) {
        if (ImGui.beginCombo(label, previewValue, flags))
            block()
        ImGui.endCombo()
    }

    inline fun combo(label: String, currentItem: KMutableProperty0<Int>, itemsSeparatedByZeros: String, heightInItems: Int = -1,
                     block: () -> Unit) {
        if (combo(label, currentItem, itemsSeparatedByZeros, heightInItems))
            block()
    }


    // Widgets: Trees

    inline fun treeNode(label: String, block: () -> Unit) {
        if (treeNode(label)) {
            block()
            treePop()
        }
    }

    inline fun treeNode(strId: String, fmt: String, block: () -> Unit) {
        if (treeNode(strId, fmt)) {
            block()
            treePop()
        }
    }

    inline fun treeNode(ptrId: Any, fmt: String, block: () -> Unit) {
        if (treeNode(ptrId, fmt)) {
            block()
            treePop()
        }
    }

    inline fun treePushed(ptrId: Any?, block: () -> Unit) {
        ImGui.treePush(ptrId)
        ImGui.treePop()
    }

    inline fun collapsingHeader(label: String, flags: TreeNodeFlags = 0, block: () -> Unit) {
        if (collapsingHeader(label, flags)) block()
    }

    inline fun collapsingHeader(label: String, open: KMutableProperty0<Boolean>, flags: TreeNodeFlags = 0, block: () -> Unit) {
        if (collapsingHeader(label, open, flags))
            block()
    }


    // Widgets: Selectables

    inline fun selectable(label: String, selected: Boolean = false, flags: Int = 0, sizeArg: Vec2 = Vec2(), block: () -> Unit) {
        if (selectable(label, selected, flags, sizeArg)) block()
    }


    // Widgets: Menus

    inline fun mainMenuBar(block: () -> Unit) {
        if (beginMainMenuBar()) {
            block()
            endMainMenuBar()
        }
    }

    inline fun menuBar(block: () -> Unit) {
        if (beginMenuBar()) {
            block()
            endMenuBar()
        }
    }

    inline fun menu(label: String, enabled: Boolean = true, block: () -> Unit) {
        if (beginMenu(label, enabled)) {
            block()
            endMenu()
        }
    }

    inline fun menuItem(label: String, shortcut: String = "", selected: Boolean = false, enabled: Boolean = true, block: () -> Unit) {
        if (menuItem(label, shortcut, selected, enabled)) block()
    }


    inline fun popupModal(name: String, pOpen: KMutableProperty0<Boolean>? = null, extraFlags: WindowFlags = 0, block: () -> Unit) {
        if (beginPopupModal(name, pOpen, extraFlags)) {
            block()
            endPopup()
        }
    }


    inline fun withTooltip(block: () -> Unit) {
        beginTooltip()
        block()
        endTooltip()
    }


    inline fun popupContextWindow(block: () -> Unit) {
        if (beginPopupContextWindow()) {
            block()
            endPopup()
        }
    }


    inline fun popup(strId: String, block: () -> Unit) {
        if (beginPopup(strId)) {
            block()
            endPopup()
        }
    }

    inline fun popupContextItem(strId: String = "", block: () -> Unit) {
        if (beginPopupContextItem(strId)) {
            block()
            endPopup()
        }
    }


}